### PR TITLE
Conditionally disable ajax add to cart 

### LINKED ISSUE #25759: Prevent ajax `adding_to_cart` conditionally
**Is your feature request related to a problem? Please describe.**

The ajax add to cart feature send an ajax request when the ajax add to cart button is clicked. I would like a way to prevent this ajax call from firing if certain conditions are not met. 

Here's the script [src](https://github.com/woocommerce/woocommerce/blob/master/assets/js/frontend/add-to-cart.js#L59-L107)

And there's a trigger in there just beforehand, but as far as I know, there's no way to hook into that trigger and prevent the code from proceeding.

```
// Trigger event.
$( document.body ).trigger( 'adding_to_cart', [ $thisbutton, data ] );
```

While generally simple products require no validation, there are times when plugins (like Name Your Price) might add fields that need validating, yet still want to take advantage of ajax add to cart.

In PHP we could use `apply_filters` to validate the product (and you do use that validating via `woocommerce_add_to_cart_validation`). I'd like to propose a JS alternative. 

**Describe the solution you'd like**
Some kind of condition/flag that will prevent the add to cart ajax call from firing. I'm proposing the same thing in One Page Checkout and it could make sense if you both used the same condition.

### PR DESCRIPTION
**Pull request by helgatheviking**
Set adding_to_cart trigger to variable. If false, do not fire ajax. Closes #25759.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:
Set `$( document.body ).triggerHandler( 'adding_to_cart', [ $thisbutton, data ] );` equal to variable. Use variable to control following `.ajax()` call.

Closes #25759 .

### How to test the changes in this Pull Request:

1. Enable ajax add to cart for shop pages
2. Add this snippet to `functions.php`
```
function kia_add_script_to_footer() {
	?>
	<script type="text/javascript">

		jQuery( function( $ ) {

			/**
			 * Validate on ajax add to cart.
			 */
			$( document.body ).on( 'adding_to_cart', function(e, $button, data ) { alert('hey');
				return false;
			});

		});

	</script>
	<?php
}
add_action( 'wp_print_footer_scripts', 'kia_add_script_to_footer' );
```
3. Go to shop archive and click add to cart. The ajax add to cart function should be disabled.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Set `adding_to_cart` trigger to variable. Allows 3rd parties to disable AJAX by returning `false`.

### CODE REVIEW

File: assets/js/frontend/add-to-cart.js [hunk 1]
```diff
@@ -76,33 +76,35 @@ jQuery( function( $ ) {
...
-					}
-
-					// Trigger event so themes can refresh other areas.
-					$( document.body ).trigger( 'added_to_cart', [ response.fragments, response.cart_hash, $thisbutton ] );
-				},
-				dataType: 'json'
-			});
+			var valid = $( document.body ).triggerHandler( 'adding_to_cart', [ $thisbutton, data ] );
+
+			if( valid !== false ) {
```
Comment (ObliviousHarmony) [inline assets/js/frontend/add-to-cart.js]: Instead of wrapping the AJAX request should we just short-circuit the function?

File: assets/js/frontend/add-to-cart.js [hunk 2]
```diff
@@ -76,33 +76,35 @@ jQuery( function( $ ) {
...
-						window.location = wc_add_to_cart_params.cart_url;
-						return;
-					}
-
-					// Trigger event so themes can refresh other areas.
-					$( document.body ).trigger( 'added_to_cart', [ response.fragments, response.cart_hash, $thisbutton ] );
-				},
-				dataType: 'json'
-			});
+			var valid = $( document.body ).triggerHandler( 'adding_to_cart', [ $thisbutton, data ] );
```
Comment (ObliviousHarmony) [inline assets/js/frontend/add-to-cart.js]: This variable's name could probably better represent its purpose.

File: assets/js/frontend/add-to-cart.js [hunk 3]
```diff
@@ -76,33 +76,35 @@ jQuery( function( $ ) {
...
+						}
+
+						// Redirect to cart option
+						if ( wc_add_to_cart_params.cart_redirect_after_add === 'yes' ) {
+							window.location = wc_add_to_cart_params.cart_url;
+							return;
+						}
+
+						// Trigger event so themes can refresh other areas.
+						$( document.body ).trigger( 'added_to_cart', [ response.fragments, response.cart_hash, $thisbutton ] );
```
Comment (ObliviousHarmony) [inline assets/js/frontend/add-to-cart.js]: When this event is not fired the button remains in the AJAX state eternally. Perhaps we should add an error event trigger as well, which can call `updateState`. We should also add a failure handler that does the same if we do, which will fix a bug here too!

**Review by ObliviousHarmony — CHANGES_REQUESTED** [review]
Thanks, this is a great start!

File: assets/js/frontend/add-to-cart.js (hunk 1 above)
Comment (helgatheviking) [reply assets/js/frontend/add-to-cart.js]: Yes, I agree with that.

File: assets/js/frontend/add-to-cart.js (hunk 2 above)
Comment (helgatheviking) [reply assets/js/frontend/add-to-cart.js]: How about `validate_ajax_adding_to_cart` ?

@ObliviousHarmony Would you take a look at what was proposed here: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/issues/694#issuecomment-597408337

Do you prefer that? Does it matter? The end result is about the same.

Comment (ObliviousHarmony) [reply assets/js/frontend/add-to-cart.js]: Perhaps `should_send_request`? While your specific use-case may be used for validation, this hook itself doesn't necessarily need to.

I'm more partial to this solution because it feels more in-line with what you'd expect of a WordPress plugin hook. Messing around with the validations API just seems unnecessary in this specific case.

File: assets/js/frontend/add-to-cart.js (hunk 3 above)
Comment (helgatheviking) [reply assets/js/frontend/add-to-cart.js]: This can be resolved by firing this _before_ the ajax state is set on the button. Or alternatively, as you say we can trigger a `failure` event and attach `update_button()` to it.

File: assets/js/frontend/add-to-cart.js (hunk 2 above)
Comment (helgatheviking) [reply assets/js/frontend/add-to-cart.js]: I didn't mean to imply we'd need the validations API, just the idea of passing an object property. But I'm fine with this way too... it does seem slightly more familiar.

Comment (helgatheviking) [reply assets/js/frontend/add-to-cart.js]: I went with `should_send_ajax_request.adding_to_cart` I don't particularly like it, but `should_send_request` by itself doesn't seem specific enough since there are lots of cases where things are sent to ajax.

File: assets/js/frontend/add-to-cart.js (hunk 3 above)
Comment (helgatheviking) [reply assets/js/frontend/add-to-cart.js]: I've added a `$( document.body ).trigger( 'ajax_request_not_sent.adding_to_cart', [ false, false, $thisbutton ] );` for when the triggerHandler is set to false. And attached `updateButton` to it.

File: assets/js/frontend/add-to-cart.js (hunk 2 above)
Comment (ObliviousHarmony) [reply assets/js/frontend/add-to-cart.js]: Sorry @helgatheviking, I meant the `valid` variable, not the event name! `adding_to_cart` was great because it was a clean non-breaking change!

File: assets/js/frontend/add-to-cart.js (hunk 3 above)
Comment (ObliviousHarmony) [reply assets/js/frontend/add-to-cart.js]: What about just `adding_to_cart_prevented`? The dot notation isn't really consistent with any other event in the cart.

**Review by claudiosanches — APPROVED** [review]
Looks good, thanks!

**Comment by helgatheviking** [comment]
@claudiosanches do you have any thoughts on @ObliviousHarmony 's comment here:
https://github.com/woocommerce/woocommerce/pull/25760/files/38e081e5907cb5bc7a3e2ea918864f7ac50aaa81#r396795350

I agree that the dot notation is weird, but I'm kinda at a loss for what to do for a naming convention.

**Comment by claudiosanches** [comment]
@helgatheviking I'm fine with the dot notation, there's other triggers on WooCommerce already that uses it.

I styles of naming for triggers that I could find in WooCommerce:

- `.` = e.g. `country-change.woocommerce`.
- `:` = e.g. `change:rates` (because Backbone)
- `-` instead of `_` = `woocommerce-product-type-change`

For just using `_` is the popular choice in WooCommerce, so if you prefer to change it will be fine for me, but I don't have anything against `.` :smiley:

**Review by ObliviousHarmony — APPROVED** [review]
